src/main/java/io/jenkins/plugins/ml/IPythonBuilder.java

```diff
@@ -75,59 +77,68 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath ws, @Nonnull Launc
             listener.getLogger().println("Executed server : " + serverName);
             // create configuration
-            IPythonUserConfig jobUserConfig = new IPythonUserConfig(serverAddress,launchTimeout,maxResults);
-            try ( IPythonInterpreterManager interpreterManager = new IPythonInterpreterManager(jobUserConfig)){
-                interpreterManager.initiateInterpreter();
-                LOGGER.info("Connection initiated successfully");
-                listener.getLogger().println("Code Parser type : " + parserType);
-                listener.getLogger().println("Executed code output : ");
-                if(parserType.equals("text")){
-                    listener.getLogger().println(interpreterManager.invokeInterpreter(code));
-                } else {
-
-                    // Run builder on selected notebook
-                    String extension = filePath.substring(filePath.lastIndexOf(".") + 1);
-                    FileExtension ext;
-                    try {
-                        // assign the extension from the enum
-                        ext = FileExtension.valueOf(extension);
-                    } catch (Exception e) {
-                        ext = FileExtension.txt;
-                    }
-                    // create file path for the file
-                    FilePath tempFilePath = ws.child(filePath);
-                    switch (ext) {
-                        case ipynb:
-                            listener.getLogger().println(StringUtils.stripStart(interpreterManager.invokeInterpreter(ConvertHelper.jupyterToText(tempFilePath)), "%text"));
-                            break;
-                        case json:
-                            // Zeppelin note book or JSON file will be interpreted line by line
-                            try (final InputStreamReader inputStreamReader = new InputStreamReader(tempFilePath.read(), Charset.forName("UTF-8"))) {
-                                Gson gson = new GsonBuilder().create();
-                                Note n = gson.fromJson(inputStreamReader, Note.class);
-                                JsonObject obj = gson.toJsonTree(n).getAsJsonObject();
-                                JsonArray array = obj.get("paragraphs").getAsJsonArray();
-                                for (JsonElement element : array)
-                                    if (element.isJsonObject()) {
-                                        // get each cell form the JSON element
-                                        JsonObject cell = element.getAsJsonObject();
-                                        String code = cell.get("text").getAsString();
-                                        listener.getLogger().println(code);
-                                        listener.getLogger().println(StringUtils.stripStart(interpreterManager.invokeInterpreter(code), "%text"));
+            launcher.getChannel().call(new MasterToSlaveCallable<Void, Throwable>() {
```

Comment:
I think we should not use `Throwable`, if we know what error we are expecting? From looking at the code, we have three possible errors in the executed block: `InterpreterException`, `InterruptedException`, and `IOException` (plus runtime errors, of course).

We can refactor the code, adding a `catch` block to the existing `try-with-resources` statement. Then decide based on the exception what to do. Maybe just wrap them in another exception.

If you prefer to leave it as-is for now @Loghijiaha , just crate a new issue so that this can be addressed later before the final release of the plugin. Examples from other plugins:

- https://github.com/jenkinsci/git-changelog-plugin/blob/ad09002e0e64d69e277eda99cc16a9abb8193da9/src/main/java/org/jenkinsci/plugins/gitchangelog/perform/RemoteCallable.java#L27
- https://github.com/jenkinsci/envinject-plugin/blob/a7e8ac38fe7be864c5b446a7a4b7ed3b6186f1d7/src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectMasterEnvVarsRetriever.java#L12